### PR TITLE
ESLint: Add `no-unmerged-classname` rule

### PR DESCRIPTION
**Pull request by mirka**
## What?

Adds a custom ESLint rule, `@wordpress/no-unmerged-classname`, and enables it for `packages/ui/src/`.

## Why?

In `@wordpress/ui`, components that spread rest props onto a JSX element can silently drop the consumer-supplied `className` if it isn't properly merged. For example:

```jsx
function Foo( { padding, ...restProps } ) {
    return <div { ...restProps } className={ styles.foo } />;
}
```

Here, any `className` passed by the consumer is in `restProps`, but the explicit `className` after the spread silently overwrites it. The correct pattern is to destructure `className` and merge it (e.g. via `clsx`).

This is a subtle mistake — the code looks reasonable and TypeScript won't flag it. A lint rule catches it at authoring time.

## How?

The rule flags any JSX element that has both an explicit `className` attribute and a spread attribute, where the enclosing function's props parameter doesn't destructure `className`. It handles:

- Object destructuring with rest element (`{ padding, ...rest }`) — flags if `className` is not among the destructured properties.
- Plain identifier params (`props`) — flags if the identifier is spread directly onto the element.
- Components that forward `className` to a different element than the one with the spread (e.g. the Tooltip pattern) — not flagged, since `className` is destructured.

The rule is not included in any preset. It's enabled for `packages/ui/src/**` only, excluding test and story files.

## Testing Instructions

```bash
npm run test:unit -- --testPathPattern='packages/eslint-plugin/rules/__tests__/no-unmerged-classname'
```

### CODE REVIEW

**Comment by mirka** [comment]
Confirmed that there are no deprecated Eslint APIs being used (#76507).

File: packages/eslint-plugin/docs/rules/no-unmerged-classname.md [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+	return <div className={ clsx( styles.foo ) } { ...restProps } />;
+}
+```
+
+```jsx
+// className destructured but not included in the className expression
+function Foo( { className, ...restProps } ) {
+	return <div className={ clsx( styles.foo ) } { ...restProps } />;
+}
+```
```
Comment (ciampo) [inline packages/eslint-plugin/docs/rules/no-unmerged-classname.md:25]: This same example is listed as valid in the tests ("className destructured but not referenced — not this rule's concern") — I feel like the rule should flag it as invalid?

File: packages/eslint-plugin/docs/rules/no-unmerged-classname.md
Comment (ciampo) [inline packages/eslint-plugin/docs/rules/no-unmerged-classname.md:1]: If I undertand the code, this rule will flag a false positive when props are destructured in the render function's body, eg

```tsx
function Foo( { ...restProps } ) {
    const { className, ...other } = restProps;
    return <div className={ clsx( styles.foo, className ) } { ...other } />;
}
```

although I guess it's rare edge case?

**Comment by ciampo** [comment]
Very nitty, but seeing https://github.com/WordPress/gutenberg/pull/76611, should we also add TS-specific tests for this rule, to make sure that it correctly handles typed parameters?

<details>

<summary>Opus suggests the following</summary>

```ts
// TypeScript-specific tests — verify the rule works with typed parameters,
// which is the norm in packages/ui/src/.
const tsRuleTester = new RuleTester( {
	parser: require.resolve( '@typescript-eslint/parser' ),
	parserOptions: {
		ecmaVersion: 2020,
		sourceType: 'module',
		ecmaFeatures: { jsx: true },
	},
} );

tsRuleTester.run( 'no-unmerged-classname (TypeScript)', rule, {
	valid: [
		{
			// Typed destructured props with className — the standard @wordpress/ui pattern
			code: `
				function Foo( { className, ...restProps }: FooProps ) {
					return <div className={ clsx( styles.foo, className ) } { ...restProps } />;
				}
			`,
		},
		{
			// forwardRef with typed props
			code: `
				const Foo = forwardRef< HTMLDivElement, FooProps >(
					function Foo( { className, ...restProps }: FooProps, ref ) {
						return <div ref={ ref } className={ clsx( styles.foo, className ) } { ...restProps } />;
					}
				);
			`,
		},
	],
	invalid: [
		{
			// Typed props without className destructured
			code: `
				function Foo( { padding, ...restProps }: FooProps ) {
					return <div className={ styles.foo } { ...restProps } />;
				}
			`,
			errors: [ { messageId: 'noUnmergedClassname' } ],
		},
		{
			// forwardRef with typed props, className not destructured
			code: `
				const Foo = forwardRef< HTMLDivElement, FooProps >(
					function Foo( { padding, ...restProps }: FooProps, ref ) {
						return <div ref={ ref } className={ styles.foo } { ...restProps } />;
					}
				);
			`,
			errors: [ { messageId: 'noUnmergedClassname' } ],
		},
	],
} );
```

</details>

File: packages/eslint-plugin/docs/rules/no-unmerged-classname.md (hunk 1 above)
Comment (mirka) [reply packages/eslint-plugin/docs/rules/no-unmerged-classname.md:25]: Good catch on the inconsistency. The docs were wrong — the rule intentionally doesn't flag this case. If the rule also checked whether `className` appears in the expression, it would false-positive on the multi-element pattern where `className` is forwarded to a different element than the one with the spread (this pattern actually exists in the package):

```jsx
function Foo( { className, style, ...props } ) {
    return (
        <Outer className={ clsx( styles.outer, className ) } style={ style }>
            <Inner className={ styles.inner } { ...props } />
        </Outer>
    );
}
```

So this is basically an intentional decision to balance complexity and practical utility. We can assume that a destructured `className` and no `no-unused-vars` error means that that `className` was used _somewhere_.

I've updated the docs to remove that example from "incorrect" and added this forwarding pattern as an explicit "correct" example (b06a60f).

File: packages/eslint-plugin/docs/rules/no-unmerged-classname.md
Comment (mirka) [reply packages/eslint-plugin/docs/rules/no-unmerged-classname.md:1]: Correct, that's a known limitation. In practice this pattern doesn't come up in `@wordpress/ui`, so I think it's fine to leave as-is for now rather than adding the complexity of tracking variable assignments through the function body.

**Comment by mirka** [comment]
> Very nitty, but seeing https://github.com/WordPress/gutenberg/pull/76611, should we also add TS-specific tests for this rule, to make sure that it correctly handles typed parameters?

I looked into this since I wasn't familiar, but short answer is no. Our rule here doesn't have any TS-specific logic. It only inspects standard JS AST nodes which both parsers produce identically regardless of type annotations (unlike the `no-dom-globals-*` rules in #76611 which handle TS-only node types like `TSTypeReference`). Adding TS tests here would be like testing for regressions in the AST parser itself, so it's unnecessary.

**Review by ciampo — APPROVED** [review]
LGTM 🚀
